### PR TITLE
Cross chain transactions

### PR DESCRIPTION
**Pull request by alfheimrShiven**

### CODE REVIEW

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    /// @return messageId The ID of the message that was sent.
+    function transferTokensPayLINK(
+        uint64 _destinationChainSelector,
+        address _receiver,
+        address _sender,
+        address _token,
+        uint256 _amount,
+        uint256 _approvedAmountLink,
+        uint256 _approvedAmountToken
+    ) external onlyOwner onlyAllowlistedChain(_destinationChainSelector) returns (bytes32 messageId) {
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: `onlyOwner` modifier? Are you planning to deploy multiple instances of this contract?

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        // Create an EVM2AnyMessage struct in memory with necessary information for sending a cross-chain message
+        //  address(linkToken) means fees are paid in LINK
+        Client.EVM2AnyMessage memory evm2AnyMessage = _buildCCIPMessage(
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol:118]: Can be converted into a memory/state variable as its being reused.

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+            _token,
+            _amount,
+            address(s_linkToken)
+        );
+
+        // Get the fee required to send the message
+        uint256 fees = s_router.getFee(_destinationChainSelector, evm2AnyMessage);
+
+        //verify amount approved for Link
+        if (_approvedAmountLink < fees) revert ApprovedAmountInsufficient(_approvedAmountLink, fees);
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: Rename error to `ApprovedLinkAmountInsufficient`

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+        //verify amount approved for token transfered
+        if (_approvedAmountToken < _amount) revert ApprovedAmountInsufficient(_approvedAmountToken, fees);
+
+        //verify
+        if (fees > s_linkToken.balanceOf(address(this)))
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: This and the condition on line 120 can be combined. Or we can get rid of the condition on line 120 as we are just checking the parameter value against the fees, rather than the actual value which is being checked in this condition using `s_linkToken.balanceOf(address(this))`

Comment (alfheimrShiven) [reply contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: Considering the Link & Transfer token approvals, their checks and withdrawals of respective tokens that are involved in this contract, we will have to deploy multiple instances of this contract, one for each account.
You can deploy and instance of this from the createAccount(….) of BaseAccountFactory.sol while setting that account address as the owner of the CCIP contract instance.

**Review by alfheimrShiven — COMMENTED** [review]
I’ve given it a read and added comments. Pls attend to them. Thanks!

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol (hunk 2 above)
Comment (codypharm) [reply contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol:118]: What are you referring to please

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol (hunk 4 above)
Comment (codypharm) [reply contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: > This and the condition on line 120 can be combined. Or we can get rid of the condition on line 120 as we are just checking the parameter value against the fees, rather than the actual value which is being checked in this condition using `s_linkToken.balanceOf(address(this))`

This checks if our contract have enough token but I guess it can be removedn since if the approved amount is enough, we can get it from the user

Comment (codypharm) [reply contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: > Considering the Link & Transfer token approvals, their checks and withdrawals of respective tokens that are involved in this contract, we will have to deploy multiple instances of this contract, one for each account. You can deploy and instance of this from the createAccount(….) of BaseAccountFactory.sol while setting that account address as the owner of the CCIP contract instance.

Am not sure I understand you

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol (hunk 2 above)
Comment (alfheimrShiven) [reply contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol:118]: The `evm2AnyMessage` variable

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol [hunk 5]
```diff
@@ -200,7 +200,7 @@ contract CrossChainTokenTransfer is OwnerIsCreator {
         address _token,
         uint256 _amount,
         uint256 _approvedAmountToken
-    ) external payable onlyOwner onlyAllowlistedChain(_destinationChainSelector) returns (bytes32 messageId) {
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: Why did you remove `onlyOwner`. That would be required right?

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol (hunk 1 above)
Comment (alfheimrShiven) [reply contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: This would be required as we plan to deploy an instance of this contract for each account.

File: contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol (hunk 4 above)
Comment (alfheimrShiven) [reply contracts/prebuilts/account/utils/CrossChainTokenTransfer.sol]: We shall discuss this over a call.

File: contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+        address _smartWalletAccount,
+        address _ccip,
+        address _link,
+        address _token,
+        uint64 _destinationChainSelector,
+        address _receiver,
+        uint _tokenAmount,
+        uint _linkAmount
+    ) public {
+        _setup_executeTransaction();
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol]: Why is smart wallet account being created here? Account should be pre-created before making a cross chain transfer request.

File: contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+    address private accountAdmin;
+
+    uint256 private accountSignerPKey = 200;
+    address private accountSigner;
+
+    uint256 private nonSignerPKey = 300;
+    address private nonSigner;
+
+    // UserOp terminology: `sender` is the smart wallet.
+    address private sender = 0xBB956D56140CA3f3060986586A2631922a4B347E;
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol]: Should be dynamic. Should be the address of the smart wallet account initiating a cross-chain request.

File: contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+contract CrossChainTokenTransferMaster {
+    // Target contracts
+    EntryPoint private entrypoint;
+    AccountFactory private accountFactory;
+
+    // Mocks
+    Number internal numberContract;
+
+    // Test params
+    uint256 private accountAdminPKey = 100;
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol]: You cannot work with static params in a production contract

File: contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+            nonce: nonce,
+            initCode: _initCode,
+            callData: _callDataForEntrypoint,
+            callGasLimit: 500_000,
+            verificationGasLimit: 500_000,
+            preVerificationGas: 500_000,
+            maxFeePerGas: 0,
+            maxPriorityFeePerGas: 0,
+            paymasterAndData: bytes(""),
+            signature: bytes("")
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol:85]: Each UserOperation object needs to have a `signature` of the account owner, which can only be generated using the account owner's `private key`, but in-case of smart accounts, we don't seem to have access to the sender's private key. You will have to find an answer to this question.

File: contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+            address(0),
+            0,
+            bytes("")
+        );
+
+        EntryPoint(entrypoint).handleOps(userOpCreateAccount, beneficiary);
+    }
+
+    /// @dev Perform many state changing transactions in a batch via Entrypoint.
+    function _initiateTokenTransferWithLink(
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol:133]: `_initiateTokenTransferWithLink (..)` will be called from the frontend right?

File: contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+            maxPriorityFeePerGas: 0,
+            paymasterAndData: bytes(""),
+            signature: bytes("")
+        });
+
+        // Sign UserOp
+        bytes32 opHash = EntryPoint(entrypoint).getUserOpHash(op);
+        bytes32 msgHash = ECDSA.toEthSignedMessageHash(opHash);
+
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(_signerPKey, msgHash);
```
Comment (alfheimrShiven) [inline contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol]: `vm.sign()` is a Foundry cheat code for writing tests and cannot be used in the contract itself. You might have to you some other external library to sign a message in a solidity contract.

File: contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol (hunk 9 above)
Comment (alfheimrShiven) [reply contracts/prebuilts/account/utils/CrossChainTokenTransferMaster.sol:85]: This logic of user ops creations will be moved to the client side, where the user will have access to their private keys, which will be used to create the `signature`
